packages/dcos-net/extra/dcos-net-watchdog.py

```diff
@@ -0,0 +1,76 @@
+#!/opt/mesosphere/bin/python
+
+import logging
+import os
+import subprocess
+import sys
+import time
+
+import dns.query
+
+
+WATCHDOG_TIMEOUT = 60
+NAME_SERVERS = ['198.51.100.1', '198.51.100.2', '198.51.100.3']
+DNS_QUERY = 'ready.spartan'
+DNS_TIMEOUT = 5
+SERVICE = 'dcos-net.service'
+
+
+def watchdog(ns):
+    try:
+        logging.info('Sending DNS query: %s', ns)
+        query = dns.message.make_query(DNS_QUERY, dns.rdatatype.ANY)
+        result = dns.query.udp(query, ns, DNS_TIMEOUT)
+        for rr in result.answer:
+            logging.info('%s', rr)
+        if not result.answer:
+            logging.error('Host not found')
+        else:
+            logging.info('%s is healthy', SERVICE)
+        return result.answer != []
+    except dns.exception.Timeout:
+        logging.error('DNS Server Timeout')
+    except:
+        logging.erorr('Exception: {}'.sys.exc_info()[1])
+    return False
+
+
+def kill(name):
+    if os.getenv('DCOS_NET_WATCHDOG') == 'false':
+        logging.info('Killing is disabled')
+        return 0
+    logging.info('Killing %s', name)
+    r = subprocess.run(['/usr/bin/env', 'systemctl', 'kill', '-s', 9, name])
```

Comment:
Perhaps we want `--kill-who=main` so that we don't kill a control process (ie., something running as `ExecStartPre=`) and incorrectly think we've killed the dcos_net service when really we've only killed bootstrap?

Reply:
Yes! It would be great not to kill bootstrap process. I'll add that in the next dcos-net PR. Thank you!

TL;DR `--kill-who=main` doesn't work properly if `KillSignal=SIGKILL` is not specified.

I wrote a small daemon to check `systemctl kill` behavior.

```
import os, time, daemon
with daemon.DaemonContext():
    os.fork()
    time.sleep(100)
```

Here are some experiments (strace will trace forked process):

```
$ systemctl kill -s 9 --kill-who=main test
Failed to kill unit test.service: No main process to kill
```

```
$ systemctl kill -s 9 --kill-who=main test
$ strace -p xxx
rt_sigaction(SIGINT, …) = 0
rt_sigaction(SIGTERM, …) = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

```
$ systemctl kill -s 9 test
$ strace -p xxx
+++ killed by SIGKILL +++
```

```
$ systemctl cat test | grep KillSignal
KillSignal=SIGKILL
$ systemctl kill --kill-who=main test
  Process: yyy ExecStart=test.py (code=exited, status=1/FAILURE)
```

```
$ systemctl cat test | grep KillSignal
KillSignal=SIGKILL
$ systemctl kill -s 9 --kill-who=main test
  Process: yyy ExecStart=test.py (code=killed, signal=KILL)
$ strace -p xxx
+++ killed by SIGKILL +++
```

Reply:
https://github.com/mesosphere/dcos/blob/urbanserj/dcos-net-env/packages/dcos-net/extra/dcos-net.service#L7
https://github.com/mesosphere/dcos/blob/urbanserj/dcos-net-env/packages/dcos-net/extra/dcos-net-watchdog.py#L43-L47